File: js/samples/05.chatModeration/src/bot.ts
import {
    Application,
    ActionPlanner,
    OpenAIModel,
    PromptManager,
    AzureOpenAIModerator,
    ModerationSeverity,
    AI,
    OpenAIModerator
} from '@microsoft/teams-ai';
import { MemoryStorage, TurnContext } from 'botbuilder';
import * as path from 'path';

if (!process.env.OPENAI_API_KEY) {
    throw new Error('Missing environment variables - please check that OPENAI_API_KEY is set.');
}

// Create AI components
const model = new OpenAIModel({
    apiKey: process.env.OPENAI_API_KEY || '',
    defaultModel: 'gpt-3.5-turbo',
    logRequests: true
});

const moderator = new OpenAIModerator({
    apiKey: process.env.OPENAI_API_KEY!,
    moderate: 'both',
});

// if (!process.env.AZURE_API_KEY) {
//     throw new Error('Missing environment variables - please check that AZURE_API_KEY is set.');
// }
//
// const model = new OpenAIModel({
//     azureApiKey: process.env.AZURE_API_KEY || '',
//     azureDefaultDeployment: 'gpt35turbo',
//     azureEndpoint: process.env.AZURE_AI_ENDPOINT || '',
//     azureApiVersion: '2023-03-15-preview',
//     logRequests: true
// });
//
// Create a moderator
// const moderator = new AzureOpenAIModerator({
//     apiKey: process.env.AZURE_MODERATOR_API_KEY!,
//     endpoint: process.env.AZURE_MODERATOR_ENDPOINT!,
//     apiVersion: '2023-04-30-preview',
//     moderate: 'both',
//     categories: [
//         {
//             category: 'Hate',
//             severity: ModerationSeverity.High
//         },
//         {
//             category: 'SelfHarm',
//             severity: ModerationSeverity.High
//         },
//         {
//             category: 'Sexual',
//             severity: ModerationSeverity.High
//         },
//         {
//             category: 'Violence',
//             severity: ModerationSeverity.High
//         }
//     ]
//     // breakByBlocklists: true,
//     // blocklistNames: [] // Text blocklist Name. Only support following characters: 0-9 A-Z a-z - . _ ~. You could attach multiple lists name here.
// });

const prompts = new PromptManager({
    promptsFolder: path.join(__dirname, '../src/prompts')
});

const planner = new ActionPlanner({
    model,
    prompts,
    defaultPrompt: 'chat',
});

// Define storage and application
const storage = new MemoryStorage();
const app = new Application({
    storage,
    ai: {
        planner,
        moderator
    }
});

// Export bots run() function
export const run = (context: TurnContext) => app.run(context);

// Listen for new members to join the conversation
app.conversationUpdate('membersAdded', async (context, state) => {
    const membersAdded = context.activity.membersAdded || [];
    for (let member = 0; member < membersAdded.length; member++) {
        // Ignore the bot joining the conversation
        // eslint-disable-next-line security/detect-object-injection
        if (membersAdded[member].id !== context.activity.recipient.id) {
            await context.sendActivity(
                `Hello and welcome! With this sample you can see the functionality of the Content Safety Moderator of Azure Open AI services.`
            );
        }
    }
});

app.message('/reset', async (context, state) => {
    state.deleteConversationState();
    await context.sendActivity(`Ok lets start this over.`);
});

app.ai.action(AI.FlaggedInputActionName, async (context, state, data) => {
    let message = '';
    if (data?.categories?.hate) {
        message += `<strong>Hate speech</strong> detected. Severity: ${data.category_scores.hate}. `;
    }
    if (data?.categories?.sexual) {
        message += `<strong>Sexual content</strong> detected. Severity: ${data.category_scores.sexual}. `;
    }
    if (data?.categories?.selfHarm) {
        message += `<strong>Self harm</strong> detected. Severity: ${data.category_scores.selfHarm}. `;
    }
    if (data?.categories?.violence) {
        message += `<strong>Violence</strong> detected. Severity: ${data.category_scores.violence}. `;
    }

    await context.sendActivity(
        `I'm sorry your message was flagged due to triggering Azure OpenAI’s content management policy. Reason: ${message}`
    );
    return AI.StopCommandName;
});

app.ai.action(AI.FlaggedOutputActionName, async (context, state, data) => {
    await context.sendActivity(`I'm not allowed to talk about such things.`);
    return AI.StopCommandName;
});

app.ai.action(AI.HttpErrorActionName, async (context, state, data) => {
    await context.sendActivity('An AI request failed. Please try again later.');
    return AI.StopCommandName;
});

File: python/samples/05.chatModeration/src/bot.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.

Description: initialize the app and listen for `message` activitys
"""

import os
import sys
import traceback

from botbuilder.core import MemoryStorage, TurnContext
from teams import Application, ApplicationOptions, TeamsAdapter
from teams.ai import AIOptions
from teams.ai.actions import ActionTurnContext, ActionTypes
from teams.ai.models import AzureOpenAIModelOptions, OpenAIModel, OpenAIModelOptions
from teams.ai.planners import ActionPlanner, ActionPlannerOptions
from teams.ai.prompts import PromptManager, PromptManagerOptions
from teams.state import TurnState
from teams.ai.moderators import Moderator, OpenAIModerator, OpenAIModeratorOptions, AzureContentSafetyModerator, AzureContentSafetyModeratorOptions

from config import Config

config = Config()

if config.OPENAI_KEY is None and config.AZURE_OPENAI_KEY is None:
    raise RuntimeError("Missing environment variables - please check that OPENAI_KEY or AZURE_OPENAI_KEY is set.")

print(config.__dict__)

model: OpenAIModel
moderator: Moderator

if config.OPENAI_KEY:
    model = OpenAIModel(
        OpenAIModelOptions(
            api_key=config.OPENAI_KEY,
            default_model="gpt-3.5-turbo"
        ))
    moderator = OpenAIModerator(
        OpenAIModeratorOptions(
            api_key=config.OPENAI_KEY,
            moderate="both",
        )   
    )
elif config.AZURE_OPENAI_KEY and config.AZURE_OPENAI_ENDPOINT:
     model = OpenAIModel(
        AzureOpenAIModelOptions(
            api_key=config.AZURE_OPENAI_KEY,
            default_model="gpt-35-turbo",
            api_version="2023-03-15-preview",
            endpoint=config.AZURE_OPENAI_ENDPOINT
        ))
     moderator = AzureContentSafetyModerator(
         AzureContentSafetyModeratorOptions(
             api_key=config.AZURE_CONTENT_SAFETY_KEY,
             moderate="both",
             api_version="2023-04-30-preview",
             endpoint=config.AZURE_CONTENT_SAFETY_ENDPOINT
         )
     )

prompts = PromptManager(
    PromptManagerOptions(
        prompts_folder=f"{os.path.dirname(os.path.abspath(__file__))}/prompts")
    )

planner = ActionPlanner(
    ActionPlannerOptions(
        model=model,
        prompts=prompts,
        default_prompt='chat'
    )
)

storage = MemoryStorage()
app = Application[TurnState](
    ApplicationOptions(
        bot_app_id=config.APP_ID,
        storage=storage,
        adapter=TeamsAdapter(config),
        ai=AIOptions(
            planner=planner,
            moderator=moderator,
        ),
    )
)

@app.message("/reset")
async def on_reset(context: TurnContext, state: TurnState):
    del state.conversation
    await context.send_activity("Ok let's start this over")
    return True

@app.ai.action(ActionTypes.FLAGGED_INPUT)
async def on_flagged_input(context: ActionTurnContext, state: TurnState):
    message = ""
    if context.data:
        data = context.data
        if data["categories"]["hate"]:
            message += f"<strong>Hate speech</strong> detected. Severity: {data['category_scores']['hate']}. "
        if data["categories"]["sexual"]:
            message += f"<strong>Sexual content</strong> detected. Severity: {data['category_scores']['sexual']}. "
        if data["categories"]["self_harm"]:
            message += f"<strong>Self harm</strong> detected. Severity: {data['category_scores']['self_harm']}. "
        if data["categories"]["violence"]:
            message += f"<strong>Violence</strong> detected. Severity: {data['category_scores']['violence']}. "
    await context.send_activity(f"I'm sorry your message was flagged due to triggering OpenAI’s content management policy. Reason: {message}")
    return ActionTypes.STOP

@app.ai.action(ActionTypes.FLAGGED_OUTPUT)
async def on_flagged_output(context: TurnContext, state: TurnState):
    await context.send_activity("I'm not allowed to talk about such things.")
    return ActionTypes.STOP

@app.ai.action(ActionTypes.HTTP_ERROR)
async def on_http_error(context: TurnContext, state: TurnState):
    await context.send_activity("An AI request failed. Please try again later")
    return ActionTypes.STOP

@app.error
async def on_error(context: TurnContext, error: Exception):
    # This check writes out errors to console log .vs. app insights.
    # NOTE: In production environment, you should consider logging this to Azure
    #       application insights.
    print(f"\n [on_turn_error] unhandled error: {error}", file=sys.stderr)
    traceback.print_exc()

    # Send a message to the user
    await context.send_activity("The bot encountered an error or bug.")
